Sources/ComplexModule/ElementaryFunctions.swift

```diff
@@ -154,78 +154,136 @@ extension Complex /*: ElementaryFunctions */ {
+    // as expected.
     //
-    // Log commutes with conjugation, so this becomes:
+    // Unfortunately, it does not handle all points close to the unit
+    // circle so well; it's easy to see why if we look at the two terms
+    // that contribute to the result. Cancellation occurs when the result
+    // is close to zero and the terms have opposing signs. By construction,
+    // the second term is always positive, so the easiest observation is
+    // that cancellation is only a problem for u < 1 (because otherwise
+    // log u is also positive, and there can be no cancellation).
+    //
+    // We are not trying for sub-ulp accuracy, just a good relative error
+    // bound, so for our purposes it suffices to have log u dominate the
+    // result:
+    if u >= 1 || u >= RealType._mulAdd(u,u,v*v) {
+      let r = v / u
+      return Complex(.log(u) + .log(onePlus: r*r)/2, θ)
+    }
+    // Here we're in the tricky case; cancellation is likely to occur.
+    // Instead of the factorization used above, we will want to evaluate
+    // log(onePlus: u² + v² - 1)/2. This all boils down to accurately
+    // evaluating u² + v² - 1. To begin, calculate both squared terms
+    // as exact head-tail products (u is guaranteed to be well scaled,
+    // v may underflow, but if it does it doesn't matter, the u term is
+    // all we need).
+    let (a,b) = Augmented.twoProdFMA(u, u)
+    let (c,d) = Augmented.twoProdFMA(v, v)
+    // It would be nice if we could simply use a - 1, but unfortunately
+    // we don't have a tight enough bound to guarantee that that expression
+    // is exact; a may be as small as 1/4, so we could lose a single bit
+    // to rounding if we did that.
+    var (s,e) = Augmented.fastTwoSum(-1, a)
+    // Now we are ready to assemble the result. If cancellation happens,
+    // then |c| > |e| > |b|, |d|, so this assembly order is safe. It's
+    // also possible that |c| and |d| are small, but if that happens then
+    // there is no significant cancellation, and the exact assembly doesn't
+    // matter.
+    s = (s + c) + e + b + d
+    return Complex(.log(onePlus: s)/2, θ)
+  }
+  
+  @inlinable
+  public static func log(onePlus z: Complex) -> Complex {
+    // log(1+z) is basically only hard to compute close to the circle
+    // |1+z| = 1; inside the circle the real part of the log is negative;
+    // outside it positive. On the circle (which passes through 0, -1±i,
+    // and -2), it is zero, so we need to be careful about cancellation.
+    //
+    // To begin, let's just see what domain z actually lies in.
+    let r2 = (1 + z).lengthSquared
+    // If |1+z| <= ½, we're well inside the circle (and close enough
+    // to -1 that 1 + z is computed exactly. If |1+z| => 2, we're far
+    // enough outside the circle that we can simply use log(1+z)--1+z
+    // may round, but that rounding doesn't cause problematic error in
+    // the final result.
+    if 2*r2 <= 1 && !(r2 < 2) { return log(1 + z) }
```

Comment:
Shouldn’t this use boolean “or” instead of “and”?

And based on the paragraph before it, those 2’s should be 4’s because of squaring, right?